packages/light-client-extension-helpers/src/smoldot/tasks.ts

```diff
@@ -0,0 +1,68 @@
+import { Client, AddChainError } from "."
+import { Effect, Fiber, Runtime, Schedule } from "effect"
+
+export const DEFAULT_SUPERVISE_REPEAT_SCHEDULE = Schedule.spaced("1 second")
+export const DEFAULT_SUPERVISE_RETRY_SCHEDULE = Schedule.spaced("1 second")
+
+export type SuperviseOptions = {
+  repeatSchedule: Schedule.Schedule<number>
+  retrySchedule: Schedule.Schedule<number>
+}
+
+/**
+ * Supervises a smoldot client by periodically invoking `addChain` with an
+ * empty chain spec.
+ *
+ * If `addChain` fails with anything other than a `AddChainError`, the client
+ * will be restarted.
+ *
+ * @param client - The smoldot client instance to be supervised.
+ * @param options - Options for customizing the supervision behavior.
+ * @param options.repeatSchedule - The frequency at which to invoke `addChain`.
+ * Defaults to {@link DEFAULT_SUPERVISE_REPEAT_SCHEDULE}.
+ * @param options.retrySchedule - The frequency at which to attempt restarting
+ * the client if needed. Defaults to {@link DEFAULT_SUPERVISE_RETRY_SCHEDULE}.
+ */
+export const supervise = (
```

Comment:
FWIW I thought I'd see what this looked like without any `Effect` stuff too and came up with this:

```ts
export type SuperviseOptions2 = {
  repeatSchedule: number, 
  retrySchedule: number,
}

export const supervise2 = async (
  client: Client,
  { repeatSchedule, retrySchedule }: SuperviseOptions2
) => {
  let stopped = false

  async function throwIfError(): Promise<boolean> {
    return client
      .addChain({ chainSpec: "" })
      .then(() => true)
      .catch((err) => {
        if (err instanceof AddChainError) {
          return true
        }
        throw err
      })
  }

  ;(async () => {
    while (!stopped) {
      try {
        if (await throwIfError()) {
          // Things are ok; wait a little and loop
          await sleep(repeatSchedule)
        }
      } catch (err) {
        // Things aren't ok; wait a little and restart.
        console.error(err)
        await sleep(retrySchedule)
        await client.restart()
      }
    }
  })()

  return {
    stop: () => {
      stopped = true
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}
```

Reply:
restart on the client needs to be supressed in try/catch but otherwise looks good :+1: